File: cpp11/Ice/async/WorkQueue.cpp
// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// **********************************************************************

#include <Ice/Ice.h>
#include <WorkQueue.h>

using namespace std;

WorkQueue::WorkQueue() : _done(false)
{
}

void
WorkQueue::start()
{
    thread t([this]()
        {
            this->run();
        });
    _thread = move(t);
}

void
WorkQueue::join()
{
    if(_thread.joinable())
    {
        _thread.join();
    }
}

void
WorkQueue::run()
{
    unique_lock<mutex> lock(_mutex);

    while(!_done)
    {
        if(_callbacks.empty())
        {
            _condition.wait(lock);
        }

        if(!_callbacks.empty())
        {
            //
            // Get next work item.
            //
            CallbackEntry entry = _callbacks.front();

            //
            // Wait for the amount of time indicated in delay to
            // emulate a process that takes a significant period of
            // time to complete.
            //
            _condition.wait_for(lock, chrono::milliseconds(get<0>(entry)));

            if(!_done)
            {
                //
                // Print greeting and send response.
                //
                _callbacks.pop_front();
                cout << "Belated Hello World!" << endl;
                auto& response = get<1>(entry);
                response();
            }
        }
    }

    //
    // Throw exception for any outstanding requests.
    //
    for(auto& entry : _callbacks)
    {
        try
        {
            throw Demo::RequestCanceledException();
        }
        catch(...)
        {
            auto& error = get<2>(entry);
            error(current_exception());
        }
    }
}

void
WorkQueue::add(int delay, function<void ()> response, function<void (exception_ptr)> error)
{
    unique_lock<mutex> lock(_mutex);

    if(!_done)
    {
        //
        // Add work item.
        //
        if(_callbacks.size() == 0)
        {
            _condition.notify_one();
        }
        _callbacks.push_back(make_tuple(delay, move(response), move(error)));
    }
    else
    {
        //
        // Destroyed, throw exception.
        //
        try
        {
            throw Demo::RequestCanceledException();
        }
        catch(...)
        {
            error(current_exception());
        }
    }
}

void
WorkQueue::destroy()
{
    unique_lock<mutex> lock(_mutex);

    //
    // Set done flag and notify.
    //
    _done = true;
    _condition.notify_one();
}

File: csharp/Ice/async/server/WorkQueue.cs
// Copyright (c) ZeroC, Inc. All rights reserved.

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

namespace Demo
{
    public class WorkQueue
    {
        private readonly List<TaskEntry> _tasks = new List<TaskEntry>();
        private bool _done;
        private Thread? _thread;

        public void Join() => _thread?.Join();

        public void Start()
        {
            _thread = new Thread(new ThreadStart(Run));
            _thread.Start();
        }

        public void Run()
        {
            lock (this)
            {
                while (!_done)
                {
                    if (_tasks.Count == 0)
                    {
                        Monitor.Wait(this);
                    }

                    if (!_done && _tasks.Count != 0)
                    {
                        //
                        // Get next work item.
                        //
                        TaskEntry entry = _tasks[0];

                        //
                        // Wait for the amount of time indicated in delay to
                        // emulate a process that takes a significant period of
                        // time to complete.
                        //
                        Monitor.Wait(this, entry.delay);

                        if (!_done)
                        {
                            _tasks.RemoveAt(0);
                            entry.task.Start();
                        }
                    }
                }

                foreach (TaskEntry e in _tasks)
                {
                    e.task.Start();
                }
            }
        }

        public Task Add(int delay)
        {
            Task t;

            lock (this)
            {
                if (!_done)
                {
                    if (_tasks.Count == 0)
                    {
                        Monitor.Pulse(this);
                    }
                    t = new Task(() => RunTask());
                    _tasks.Add(new TaskEntry(t, delay));
                }
                else
                {
                    t = Task.Factory.StartNew(() => RunTask());
                }
            }

            return t;
        }

        public void Destroy()
        {
            lock (this)
            {
                _done = true;
                Monitor.Pulse(this);
            }
        }

        private void RunTask()
        {
            lock (this)
            {
                if (!_done)
                {
                    Console.Out.WriteLine("Belated Hello World!");
                }
                else
                {
                    throw new RequestCanceledException();
                }
            }
        }

        private class TaskEntry
        {
            public TaskEntry(Task task, int delay)
            {
                this.task = task;
                this.delay = delay;
            }

            public Task task;
            public int delay;
        }
    }

}
